server/app/assets/javascripts/session.ts

```diff
@@ -0,0 +1,396 @@
+import {ToastController} from './toast'
+
+/**
+ * Represents session timeout data with timestamps for various timeout events.
+ * All timestamps are Unix timestamps in seconds.
+ */
+interface TimeoutData {
+  /** When to show the inactivity warning modal */
+  inactivityWarning: number
+  /** When to logout due to inactivity */
+  inactivityTimeout: number
+  /** When to show the total session length warning modal */
+  totalWarning: number
+  /** When to logout due to total session length */
+  totalTimeout: number
+  /** Server's current time when cookie was set - used for clock skew calculation */
+  currentTime: number
+}
+
+/**
+ * Types of warning modals that can be shown to the user.
+ */
+export enum WarningType {
+  /** Warning shown when user has been inactive */
+  INACTIVITY = 'inactivity',
+  /** Warning shown when total session length is approaching limit */
+  TOTAL_LENGTH = 'total_length',
+}
+
+/**
+ * Handles session timeout management including:
+ * - Reading timeout data from cookies
+ * - Showing warning modals before timeout
+ * - Handling session extension requests
+ * - Managing automatic logout
+ */
+export class SessionTimeoutHandler {
+  /** Name of cookie storing timeout data */
+  private static COOKIE_NAME = 'session_timeout_data'
+  /** Tracks if inactivity warning is currently shown */
+  private static inactivityWarningShown = false
+  /** Tracks if total length warning is currently shown */
+  private static totalLengthWarningShown = false
+  /** Tracks if inactivity warning has been shown at least once */
+  private static hasInactivityWarningBeenShown = false
+  /** Tracks if total length warning has been shown at least once */
+  private static hasTotalLengthWarningBeenShown = false
+  /** Current active timeout timer */
+  private static timer: number | null = null
+
+  static init() {
+    void this.checkAndSetTimer()
```

Comment:
nitpick (non-blocking): It would be helpful to have the `checkAndSetTimer` method defined directly under this the `init` method for easy reference.

Reply:
Make sense to have these methods closer.